bgbb/sql/sql_utils.py

```diff
@@ -0,0 +1,219 @@
+import datetime as dt
+from typing import List
+
+import pandas as pd
+
+from bgbb.sql.sql_utils_tests import test_ho_range, test_model_range, test_rft
+
+
+def to_s3_fmt(date): return date.strftime('%Y%m%d')
+
+
+def to_samp_ids(samp_ids: List[int]) -> str:
+    """
+    iter of ints to SQL string version for main_summary
+    >>> to_samp_ids([0, 1, 2])
+    "'0', '1', '2'"
+    """
+    mn = min(samp_ids)
+    mx = max(samp_ids)
+    if mn == 1:
+        print('Warning: 0 is minimum sample_id allowed, not 1')
+    elif mn < 0:
+        raise ValueError('sample_id >= "0" in main summary')
+    if mx > 99:
+        raise ValueError('sample_id <= "99" in main summary')
+    return to_sql_list(map(str, samp_ids))
+    # (', '.join("'{}'".format(i) for i in samp_ids))
+
+
+def to_sql_list(xs):
+    """stringify lists for SQL queries
+    >>> to_sql_list([1, 2, 3]) == '1, 2, 3'
+    """
+    def to_sql_literal(x):
+        if isinstance(x, str):
+            return "'{}'".format(x)
+        return str(x)
+    res = ', '.join(map(to_sql_literal, xs))
+    return res
+
+
+def test_to_sql_list():
+    assert to_sql_list(['GB']) == "'GB'"
+    assert to_sql_list(['GB', 'US', 'IN']) == "'GB', 'US', 'IN'"
+    assert to_sql_list([1, 2, 3]) == '1, 2, 3'
+
+
+def insert_country(q, insert_before="{sample_comment}",
+                   countries: List[str]=["GB"]):
+    "Insert country restriction into SQL string for testing"
+    i = q.find(insert_before)
+    to_insert = "AND country IN ({})\n      ".format(to_sql_list(countries))
+    return q[:i] + to_insert + q[i:]
+
+
+def mk_time_params(HO_WIN=14, MODEL_WIN=90, ho_start='2018-08-01'):
+    """
+    Return container whose attributes are holdout and model input
+    date ranges, specified by a training window `MODEL_WIN`,
+    holdout evaluation window `HO_WIN` and holdout start date `ho_start`
+    (day after last day in model window).
+    """
+    def r():
```

Comment:
This is interesting, but might be clearer to build up a dictionary and passing them around as keyword arguments.